src/java/org/apache/cassandra/utils/SystemInfo.java

```diff
@@ -0,0 +1,223 @@
+        if (oshi.SystemInfo.getCurrentPlatform() == PlatformEnum.LINUX)
+        {
+            String path = format("/proc/%s/limits", getPid());
+            try
+            {
+                List<String> lines = FileUtils.readLines(new File(path));
+                for (String line : lines)
+                {
+                    if (line.startsWith("Max processes"))
+                    {
+                        String[] parts = SPACES_PATTERN.split(line);
+
+                        if (parts.length < 3)
+                            continue;
+
+                        String limit = parts[2];
+                        return "unlimited".equals(limit) ? INFINITY : Long.parseLong(limit);
+                    }
+                }
+                logger.error( "'Max processes' not found in "+path);
+            }
+            catch (Throwable t)
+            {
+                logger.error(format("Unable to read %s", path), t);
+            }
+        }
+
+        /* return the default value for non-Linux systems or parsing error.
+         * Can not return 0 as we know there is at least 1 process (this one) and
+         * -1 historically represents infinity.
+         */
+        return DEFAULT_MAX_PROCESSES;
+    }
+
+    /**
+     * @return The maximum number of open files allowd to the current process/user.
+     */
+    public long getMaxOpenFiles()
+    {
+        // ulimit -H -n
+        return si.getOperatingSystem().getCurrentProcess().getHardOpenFileLimit();
+    }
+
+    /**
+     * Gets the Virtual Memory Size (VSZ). Includes all memory that the process can access,
+     * including memory that is swapped out and memory that is from shared libraries.
+     *
+     * @return The amount of virtual memory allowed to be allocatedby the current process/user.
+     */
+    public long getVirtualMemoryMax()
+    {
+        return si.getOperatingSystem().getCurrentProcess().getVirtualSize();
+    }
+
+    /**
+     * @return The amount of swap space allocated on the system.
+     */
+    public long getSwapSize()
+    {
+        return si.getHardware().getMemory().getVirtualMemory().getSwapTotal();
+    }
+
+    /**
+     * @return the PID of the current system.
+     */
+    public long getPid()
+    {
+        return si.getOperatingSystem().getProcessId();
+    }
+
+    /**
+     * @return the Semver for the kernel version of the OS.
+     */
+    public Semver getKernelVersion()
+    { 
+        return new Semver(si.getOperatingSystem().getVersionInfo().getBuildNumber(), Semver.SemverType.LOOSE);
+    }
+
+    /**
+     * Tests if the system is running in degraded mode.
+     *
+     * @return non-empty optional with degradation messages if the system is in degraded mode, empty optional otherwise.
+     */
+    public Optional<String> isDegraded()
+    {
+        Supplier<String> expectedNumProc = () -> {
+            // only check proc on nproc linux
+            if (oshi.SystemInfo.getCurrentPlatform() == PlatformEnum.LINUX)
```

Comment:
@Claudenw we can use `platform` method of this class instead of `oshi.SystemInfo.getCurrentPlatform()` and `platform` might return `PlatformEnum` instead of `String`.

That way, except reusing the code, it will also unblocks us when we will go to mock what this class is doing when testing various scenarios.

Reply:
@Claudenw still, use that `platform()` method here and make it return String (as it is now) and just call `Platform.LINUX.name()` here and compare it with `equals`.

Reply:
@Claudenw one nice advantage of returning enum is that you can use it in `case` and it is compile time safe ... We can just compare enums directly. I do not think it is a lot of ask for given how comfortable its usage will be. We are not planning to change the impl anytime soon again.

Reply:
@Claudenw do you plan to use `platform()` here?